scripts/bert/pretraining_utils.py

```diff
@@ -0,0 +1,288 @@
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""Utilities for pre-training."""
+import glob
+import time
+import os
+import logging
+import argparse
+import mxnet as mx
+from mxnet.gluon.data import DataLoader
+
+import gluonnlp as nlp
+from gluonnlp.data.batchify import Tuple, Stack, Pad
+from gluonnlp.metric import MaskedAccuracy
+
+__all__ = ['get_model_loss', 'get_pretrain_dataset', 'get_dummy_dataloader',
+           'save_params', 'evaluate', 'forward', 'split_and_load', 'get_argparser']
+
+def get_model_loss(ctx, model, pretrained, dataset_name, dtype, ckpt_dir=None, start_step=None):
+    """Get model for pre-training."""
+    # model
+    model, vocabulary = nlp.model.get_model(model,
+                                            dataset_name=dataset_name,
+                                            pretrained=pretrained, ctx=ctx)
+
+    if not pretrained:
+        model.initialize(init=mx.init.Normal(0.02), ctx=ctx)
+    model.cast(dtype)
+
+    if ckpt_dir and start_step:
+        param_path = os.path.join(ckpt_dir, '%07d.params'%start_step)
+        model.load_parameters(param_path, ctx=ctx)
+        logging.info('Loading step %d checkpoints from %s.', start_step, param_path)
+
+    model.hybridize(static_alloc=True)
+
+    # losses
+    nsp_loss = mx.gluon.loss.SoftmaxCELoss()
+    mlm_loss = mx.gluon.loss.SoftmaxCELoss()
+    nsp_loss.hybridize(static_alloc=True)
+    mlm_loss.hybridize(static_alloc=True)
+
+    return model, nsp_loss, mlm_loss, vocabulary
+
+def get_pretrain_dataset(data, batch_size, num_ctxes, shuffle, use_avg_len,
+                         num_buckets, num_parts=1, part_idx=0, prefetch=True):
+    """create dataset for pretraining."""
+    num_files = len(glob.glob(os.path.expanduser(data)))
+    logging.debug('%d files found.', num_files)
+    assert num_files >= num_parts, \
+        'Number of training files must be greater than the number of partitions'
+    split_sampler = nlp.data.SplitSampler(num_files, num_parts=num_parts, part_index=part_idx)
+    stream = nlp.data.SimpleDatasetStream(nlp.data.NumpyDataset, data, split_sampler)
+    if prefetch:
+        stream = nlp.data.PrefetchingStream(stream)
+
+    def get_dataloader(dataset):
+        """create data loader based on the dataset chunk"""
+        lengths = dataset.get_field('valid_lengths')
+        # A batch includes: input_id, masked_id, masked_position, masked_weight,
+        #                   next_sentence_label, segment_id, valid_length
+        batchify_fn = Tuple(Pad(), Pad(), Pad(), Pad(), Stack(), Pad(), Stack())
```

Comment:
The default pad_val of `Pad()` is 0, but is the pad_val of each field 0?

Reply:
Good catch. I think it's a general issue for gluon-nlp. For this script, the model forward and loss computation relies on the mask_id and valid_length of the data batch. Pad does not affect the correctness. Let me create an issue for this. I think we can print a warning `using default padding value 0` if user does not specify the pad value when using `Pad()`